torch/distributed/_composable/fully_shard.py

```diff
@@ -81,4 +83,6 @@ def fully_shard(
     _register_pre_forward_hooks(state, modules)
     _register_post_forward_hooks(state, modules)
     _register_root_pre_forward_hook(state, module)  # prepend last
+    for submodule in state._module_to_handles.keys():
```

Comment:
I am not sure if this `for` loop is exactly what we want. This is probably my fault for making `_module_to_handles` confusing. The following is going to get into some details about FSDP.

The keys of the `dict` are computed as the union over all `handle._modules` for all `handles` in `state._handles`. However, `handle._modules` only includes the modules that have parameters in the `handle.flat_param`. 

For example, if you have
```
parent(
    child()
    child()
)
``` 
where only each `child` has parameters but `parent` does not, then `parent` will not be in `handle._modules`. In that case, IIUC, `parent` is not given access to the `state`. In other words, the invariant here is that, if we apply `fully_shard(module)`, then a submodule in `module`'s tree is given access to `state` iff the submodule has parameters.

Perhaps, this is _sufficient_ for your use case in checkpointing? Regardless, I can see that this might be confusing.

One way to fix this depends on one of my not-yet-landed PRs: https://github.com/pytorch/pytorch/pull/90387. That PR will introduce a `comm_module` for each `FlatParamHandle`, which for the wrapper code path is exactly the same as the `module` passed to `FlatParamHandle.__init__()` and for the composable path represents that same module. The difference is that for the composable `fully_shard` path, we always pass the local FSDP root module as `module` to `FlatParamHandle.__init__()`, which may not be the same module we would have passed into `__init__()` for the wrapper path. `comm_module` handles that difference.

If we have `comm_module`, then we can do something like:
```
all_submodules = set()
for handle in _all_handles(state):
    for submodule in handle._comm_module.modules():
        all_submodules.add(submodule)
for submodule in all_submodules:
    _insert_module_state(submodule, state)
```
Note that the above snippet has a quadratic complexity with the number of submodules, which is not ideal, but should be okay for now. The reason is that `_comm_module.modules()` will "double-count" modules if one `_comm_module` is the parent of another.

If the way you have it in this PR is okay for now, then we do not need to change this one. We can wait for my `comm_module` PR to land and then adjust after.

This is kind of convoluted, so please let me know if anything is unclear!

Reply:
Discussed offline: We can just do something like:
```
for submodule in module.modules():
    if submodule not in state._ignored_modules:
        _insert_module_state(submodule, state)
```
(without relying on `_module_to_handles`)